app/src/main/java/com/example/photostomusic/fragments/RecommendationFragment.java

```diff
@@ -249,27 +254,53 @@ public void discarded(int mIndex, int direction) {
                                 case 0:
                                     // Song rejected, it is now removed from stack
                                     Log.d(TAG, "discarded: 0");
+                                    mp.stop();
+                                    mp.reset();
                                     break;
                                 case 1:
                                     // Song accepted, sending to Parse
                                     Log.d(TAG, "discarded: 1");
-                                    saveSong(songImage, songName, songArtist, songAlbum, songCode, genres, photo);
+                                    try {
+                                        saveSong(songs.getJSONObject(mIndex-1), genres, photo);
+                                    } catch (JSONException e) {
+                                        e.printStackTrace();
+                                    }
                                     break;
                                 case 2:
                                     // Song rejected, it is now removed from stack
                                     Log.d(TAG, "discarded: 2");
+                                    mp.stop();
+                                    mp.reset();
                                     break;
                                 case 3:
                                     // Accept, send to parse and move to history
                                     Log.d(TAG, "discarded: 3");
-                                    saveSong(songImage, songName, songArtist, songAlbum, songCode, genres, photo);
+                                    try {
+                                        saveSong(songs.getJSONObject(mIndex), genres, photo);
+                                    } catch (JSONException e) {
+                                        e.printStackTrace();
+                                    }
```

Comment:
I would rearrange your switch statement to minimize code duplication. so something more like
```
case 1:
case 3:
    try {
        saveSong(songs.getJSONObject(mIndex-1), genres, photo);
    } catch (JSONException e) {
        e.printStackTrace();
    }
    return;
case 0:
case 2:
    Log.d(TAG, "Song discarded");
default:
    mp.stop();
    mp.reset();
```

Reply:
I had no idea this was possible! It has now been changed :)
Thanks for the tip!